### PR TITLE
Add substitution with options

### PR DESCRIPTION
**Pull request by srebhan**
This PR introduces the `SubstituteWithOptions()` function including options for custom patterns, custom substitution functions and a **new** replacement function for allowing to customize the replacement behavior if e.g. no mapping was found.
The existing `SubstituteWith` function is switched over to the newly introduced function.

### CODE REVIEW

**Comment by ndeloof** [comment]
as there's no issue associated with this PR, can you please clarify how you expect this to be used ? Any upstream project relying on this ?

**Comment by srebhan** [comment]
@ndeloof we are using this library for [Telegraf](https://github.com/influxdata/telegraf) and have to reproduce a specific behavior namely to only substitute environment-variable expressions if those environment variables are defined. This requirement comes from keeping backward compatibility to not break existing configurations. The issue linked above shows how we use this new function in code.

The intension is to be able to better control the replacement process while keeping the nice shell-like interpolation functionality.

Please let me know if I should open a corresponding issue!

**Comment by ndeloof** [comment]
For legal reasons we require commits to be signed-off, please amend your commit and force push

**Review by ndeloof — APPROVED** [review]

**Review by glours — COMMENTED** [review]
Is it possible to add an unit test checking the correct behavior of `SubstituteWithOptions` with a replacement function, please?

File: template/template_test.go [hunk 1]
```diff
@@ -377,9 +377,33 @@ func TestSubstituteWithCustomFunc(t *testing.T) {
...
+	result, err := SubstituteWithOptions("ok ${FOO}", defaultMapping, options...)
+	assert.NilError(t, err)
+	assert.Check(t, is.Equal("ok first", result))
+
+	result, err = SubstituteWithOptions("ok ${BAR}", defaultMapping, options...)
+	assert.NilError(t, err)
+	assert.Check(t, is.Equal("ok foobar", result))
+
+	_, err = SubstituteWithOptions("ok ${NOTHERE}", defaultMapping, options...)
+	assert.Check(t, is.ErrorContains(err, "bad choice dude"))
```
Comment (glours) [inline template/template_test.go:402]: Removing the `dude` in the error message, it doesn't provide any value to your test 😉 
```suggestion
				return "", fmt.Errorf("bad choice: %q", s)
			}
			r, err := DefaultReplacementFunc(s, m, c)
			if err == nil && r != "" {
				return r, nil
			}
			return "foobar", nil
		}),
	}
	result, err := SubstituteWithOptions("ok ${FOO}", defaultMapping, options...)
	assert.NilError(t, err)
	assert.Check(t, is.Equal("ok first", result))

	result, err = SubstituteWithOptions("ok ${BAR}", defaultMapping, options...)
	assert.NilError(t, err)
	assert.Check(t, is.Equal("ok foobar", result))

	_, err = SubstituteWithOptions("ok ${NOTHERE}", defaultMapping, options...)
	assert.Check(t, is.ErrorContains(err, "bad choice"))
```

**Review by glours — COMMENTED** [review]
Just a small modification in your test and we're ready to merge the PR

**Review by glours — APPROVED** [review]

**Pull request merged**